### PR TITLE
MOTECH-2504: Improved logging of the HTTP-Agent module

### PR DESCRIPTION
**Pull request by mgrabowski7**

### CODE REVIEW

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by sebbrudzinski** [comment]
test this please

File: http-agent/pom.xml [hunk 1]
```diff
@@ -43,6 +43,12 @@
             <groupId>org.apache.httpcomponents</groupId>
             <artifactId>httpcore-osgi</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
```
Comment (pgesek) [inline http-agent/pom.xml]: We do not want a different version of spring-core for tests

File: http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import org.apache.commons.lang.builder.EqualsBuilder;
+import org.apache.commons.lang.builder.HashCodeBuilder;
+import org.motechproject.mds.annotations.Entity;
+import org.motechproject.mds.annotations.Field;
+
+/**
+ * The <code>HTTPActionRecord</code> class represents a record of a http Action.
+ */
+@Entity
+public class HTTPActionRecord {
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java]: It says HTTPActionAudit in the ticket

File: http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    private String responseCode;
+
+    public HTTPActionRecord(String url, String request, String responseBody, String responseCode) {
+        this.url = url;
+        this.request = request;
+        this.responseBody = responseBody;
+        this.responseCode = responseCode;
+    }
+
+    public void setUrl()  { this.url = url; }
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java]: Place the content of the method in the new line. Same with closing brace. http://docs.motechproject.org/en/latest/development/coding_conventions.html#braces

File: http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    public void setRequesrt()  { this.request = request; }
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java]: Place the content of the method in the new line. Same with closing brace. http://docs.motechproject.org/en/latest/development/coding_conventions.html#braces

File: http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    public void setResponseBody()  { this.responseBody = responseBody; }
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java]: Place the content of the method in the new line. Same with closing brace. http://docs.motechproject.org/en/latest/development/coding_conventions.html#braces

File: http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+    public void setResponseCode()  { this.responseCode = responseCode; }
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java]: Place the content of the method in the new line. Same with closing brace. http://docs.motechproject.org/en/latest/development/coding_conventions.html#braces

File: http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    public String getUrl() {
+        return url;
+    }
+
+    public String getResponseCode() {
+        return responseCode;
+    }
+
+    public String getResponseBody() {
+
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java]: Remove empty line.

File: http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        return responseBody;
+    }
+
+    public String getRequest() {
+
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java]: Remove empty line.

File: http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java (hunk 4 above)
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java]: This setter has got a typo. (Don't you auto-generate setters/getters?)

File: http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        return new EqualsBuilder()
+                .append(url, that.url)
+                .append(request, that.request)
+                .append(responseBody, that.responseBody)
+                .append(responseCode, that.responseCode)
+                .isEquals();
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionRecord.java]: Why not use Objects.equals and Objects.hash from Java7?

File: http-agent/src/test/java/org/motechproject/http/agent/listener/HttpClientEventListenerTest.java [hunk 10]
```diff
@@ -7,23 +7,29 @@
...
 import org.motechproject.event.MotechEvent;
 import org.motechproject.http.agent.domain.Credentials;
 import org.motechproject.http.agent.domain.EventDataKeys;
 import org.motechproject.http.agent.domain.EventSubjects;
+import org.motechproject.http.agent.domain.HTTPActionRecord;
+import org.motechproject.http.agent.service.HTTPActionService;
 import org.motechproject.http.agent.service.Method;
 import org.motechproject.config.SettingsFacade;
-import org.springframework.http.HttpEntity;
+import org.springframework.http.*;
```
Comment (sebbrudzinski) [inline http-agent/src/test/java/org/motechproject/http/agent/listener/HttpClientEventListenerTest.java]: Avoid star imports

File: http-agent/src/test/java/org/motechproject/http/agent/listener/HttpClientEventListenerTest.java [hunk 11]
```diff
@@ -37,27 +43,41 @@
...
+    HTTPActionRecord httpActionRecord;
 
-    private HttpClientEventListener httpClientEventListener;
 
+    private HttpClientEventListener httpClientEventListener;
     private String url;
     private String data;
     private Map headers;
     private Credentials credentials;
+    private HttpEntity httpEntity;
```
Comment (sebbrudzinski) [inline http-agent/src/test/java/org/motechproject/http/agent/listener/HttpClientEventListenerTest.java]: Any reason httpEntity and request are class fields? It seems like they can be method variables.

File: http-agent/src/test/java/org/motechproject/http/agent/listener/HttpClientEventListenerTest.java [hunk 12]
```diff
@@ -68,12 +88,17 @@ public void shouldReadFromQueueAndMakeAHttpCallForPost() throws IOException {
...
+        httpHeaders.add("key2", "value2");
+        request = new HttpEntity<String>(data,httpHeaders);
+        responseEntity = new ResponseEntity<String>(body,HttpStatus.OK);
+        when(restTempate.exchange(eq(url),eq(HttpMethod.POST),eq(request), eq(String.class))).thenReturn((ResponseEntity<String>) responseEntity);
         httpClientEventListener.handle(motechEvent);
 
-        ArgumentCaptor<Object> captor = ArgumentCaptor.forClass(Object.class);
-        verify(restTempate).postForLocation(eq(url), captor.capture());
-        assertRequestObject(data, headers, captor.getValue());
+        HTTPActionRecord httpActionRecord = new HTTPActionRecord(url, request.toString(), responseEntity.getBody().toString(),
```
Comment (sebbrudzinski) [inline http-agent/src/test/java/org/motechproject/http/agent/listener/HttpClientEventListenerTest.java]: We should have a separate test that checks logging.

**Comment by sebbrudzinski** [comment]
What about HttpAgentImpl? it seems like logging should happen there as well.

File: http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionAudit.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+import org.motechproject.mds.annotations.Entity;
+import org.motechproject.mds.annotations.Field;
+
+import java.util.Objects;
+
+/**
+ * The <code>HTTPActionAudit</code> class represents a record of a http Action.
+ */
+@Entity
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/domain/HTTPActionAudit.java]: Let's set it nonEditable, just like other audit entities in MOTECH.

File: http-agent/src/test/java/org/motechproject/http/agent/listener/HttpClientEventListenerTest.java [hunk 14]
```diff
@@ -100,16 +118,35 @@ public void shouldReadFromQueueAndMakeAHttpDeleteCall() throws IOException {
...
+        httpHeaders.add("key2", "value2");
+        request = new HttpEntity<String>(data,httpHeaders);
+        responseEntity = new ResponseEntity<String>(body,HttpStatus.OK);
+        when(restTempate.exchange(eq(url),eq(HttpMethod.DELETE),eq(request), eq(String.class))).thenReturn((ResponseEntity<String>) responseEntity);
+
+        httpClientEventListener.handle(motechEvent);
+
+        HTTPActionAudit httpActionAudit = new HTTPActionAudit(url, request.toString(), responseEntity.getBody().toString(),
+                responseEntity.getStatusCode().toString());
+        verify(httpActionService,times(1)).create(eq(httpActionAudit));
```
Comment (sebbrudzinski) [inline http-agent/src/test/java/org/motechproject/http/agent/listener/HttpClientEventListenerTest.java]: one execution is the default when calling verify with one argument

**Comment by sebbrudzinski** [comment]
test this please

**Comment by sebbrudzinski** [comment]
@pgesek OK to merge

**Pull request merged**